voip_utils/sip.py

```diff
@@ -176,106 +334,258 @@ def connection_made(self, transport):
-                + r"(?:\?(?P<headers>.*))?"  # headers
-            )
-            re_uri = re_splituri.search(ruri)
-            if re_uri is None:
-                raise ValueError("Receiver URI did not match expected pattern")
-
-            server_ip = re_uri.group("host")
-            if not is_ipv4_address(server_ip):
-                raise VoipError(f"Invalid IPv4 address in {ruri}")
-
-            self.on_call(
-                CallInfo(
-                    caller_endpoint=caller_endpoint,
-                    caller_rtp_port=caller_rtp_port,
-                    server_ip=server_ip,
-                    headers=headers,
-                    opus_payload_type=opus_payload_type,
+                # Acknowledge the BYE message, otherwise the phone will keep sending it
+                rtp_info = get_rtp_info(smsg.body)
+                remote_rtp_ip = rtp_info.rtp_ip
+                remote_rtp_port = rtp_info.rtp_port
+                opus_payload_type = rtp_info.payload_type
+                # We should remove the call from the outgoing calls dict now if it is there
+                self._end_outgoing_call(smsg.headers["call-id"])
+                ok_lines = [
+                    "SIP/2.0 200 OK",
+                    f"Via: {smsg.headers['via']}",
+                    f"From: {smsg.headers['from']}",
+                    f"To: {smsg.headers['to']}",
+                    f"Call-ID: {smsg.headers['call-id']}",
+                    f"CSeq: {smsg.headers['cseq']}",
+                    f"User-Agent: {VOIP_UTILS_AGENT} 1.0",
+                    "Content-Length: 0",
+                ]
+                ok_text = _CRLF.join(ok_lines) + _CRLF
+                ok_bytes = ok_text.encode("utf-8")
+                # We should probably tell the associated RTP server to shutdown at this point, assuming we aren't reusing it for other calls
+                _LOGGER.debug("Sending OK for BYE message: %s", ok_text)
+                self.transport.sendto(
+                    ok_bytes,
+                    (caller_ip, caller_sip_port),
                 )
-            )
+                # The transport might be used for incoming calls
+                # as well, so we should leave it open.
+
+                # Cleanup any necessary call state
+                self.on_hangup(
+                    CallInfo(
+                        caller_endpoint=caller_endpoint,
+                        local_endpoint=local_endpoint,
+                        caller_rtp_port=remote_rtp_port,
+                        server_ip=remote_rtp_ip,
+                        headers=smsg.headers,
+                    )
+                )
+
         except Exception:
-            _LOGGER.exception("Unexpected error handling SIP INVITE")
+            _LOGGER.exception("Unexpected error handling SIP message")
 
     @abstractmethod
     def on_call(self, call_info: CallInfo):
         """Handle incoming calls."""
 
+    def on_hangup(self, call_info: CallInfo):
```

Comment:
I don't think this method needs to be required to be implemented in subclasses. That way if we want to release and use the updated voip_utils without updating HA yet things should still work.